File: brick/src/main/java/com/numeon/brick/AbstractViewModel.java
package com.numeon.brick;

import androidx.lifecycle.ViewModel;

import org.jetbrains.annotations.NotNull;

public abstract class AbstractViewModel<V extends IView, M extends IModel> extends ViewModel implements IPresenter<V, M> {

    private V view;
    private M model;

    @NotNull
    @Override
    final public V getView() {
        return view;
    }

    @NotNull
    @Override
    final public M getModel() {
        return model;
    }

    @Override
    final public void onCreated(V view) {
        if (this.view == null) {
            this.view = view;
        }
        if (this.model == null) {
            this.model = ModelFactory.create(this);
        }
    }

}

File: brick/src/main/java/com/numeon/brick/AbstractViewModel.kt
package com.numeon.brick

import androidx.lifecycle.ViewModel

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.MainScope

abstract class AbstractViewModel<V : IView, M : IModel> : ViewModel(), IPresenter<V, M>, CoroutineScope by MainScope() {

    private lateinit var view: V
    private lateinit var model: M

    final override fun getView(): V {
        return view
    }

    final override fun getModel(): M {
        return model
    }

    fun attachView(view: V) {
        if (!this::view.isInitialized) {
            this.view = view
        }
        if (!this::model.isInitialized) {
            this.model = ModelFactory.create(this)
        }
    }

}
